From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sat, 22 Jul 2023 21:00:42 +0200
Subject: [PATCH] Client perspective - Item replacements - Perform replacements
 in FriendlyByteBuf

License: Fiddle Public License Agreement 1.0 (included in license/FPL-1.0.txt)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/network/FriendlyByteBuf.java b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
index 321aff31e6d19e39fd4d442d9e2b423360c92adf..a83826f6891aa723e3168805c42414e077a688ad 100644
--- a/src/main/java/net/minecraft/network/FriendlyByteBuf.java
+++ b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
@@ -84,6 +84,8 @@ import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
 import org.fiddlemc.fiddle.packet.ClientPerspective;
 import org.fiddlemc.fiddle.packet.PacketWithExplicitTargetClientPerspective;
+import org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext;
+import org.fiddlemc.fiddle.packet.item.ItemStackInPacketReplacer;
 import org.jetbrains.annotations.NotNull;
 import org.joml.Quaternionf;
 import org.joml.Vector3f;
@@ -688,10 +690,18 @@ public class FriendlyByteBuf extends ByteBuf {
     }
 
     public FriendlyByteBuf writeItem(ItemStack stack) {
+        // Fiddle start - client perspective - item replacements - FriendlyByteBuf
+        // TODO instead of calling this method, call with the correct context for non-DEFAULT ones (stonecutter, in item frame)
+        return this.writeItem(stack, ItemStackInPacketContext.DEFAULT);
+    }
+
+    public FriendlyByteBuf writeItem(ItemStack stack, @NotNull ItemStackInPacketContext context) {
+        // Fiddle end - client perspective - item replacements - FriendlyByteBuf
         if (stack.isEmpty() || stack.getItem() == null) { // CraftBukkit - NPE fix itemstack.getItem()
             this.writeBoolean(false);
         } else {
             this.writeBoolean(true);
+            stack = ItemStackInPacketReplacer.replace(stack, context, this.getExplicitTargetClientPerspective(), this.getExplicitTargetLocale()); // Fiddle - client perspective - item replacements - FriendlyByteBuf
             Item item = stack.getItem();
 
             this.writeId(BuiltInRegistries.ITEM, item);
